### ISSUE TITLE
[F#] Implement new Concept Exercise: classes

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#1015]
This issue describes how to implement the `classes` concept exercise for the F# track.

## Goal

The goal of this exercise is to teach the student the basics of the Concept of Classes in F#.

## Learning objectives

- Know what classes are.
- Know what the difference is between classes and records and when to use which.
- Know what encapsulation is.
- Know what composition is.
- Know what fields are.
- Know how to define a primary and additional constructors.
- Know how to use `do` bindings in classes.
- Know how to create an object.
- Know how to define methods.
- Know how to update state through methods.
- Know how to define `let` and `val` bindings in a class.
- Know what static bindings are and how to define them.
- Know about the `unit` type.
- Know about the `public` and `private` visibility modifiers.

## Out of scope

- Interfaces.
- Inheritance.
- Properties.
- Indexers.
- Structs.
- Method overloading.

## Concepts

The Concepts this exercise unlocks are:

- `classes-basic`: know what classes are; know what the difference is between classes and records and when to use which; know what encapsulation is; know what composition is; know what fields are; know how to define a primary and additional constructors; know how to use `do` bindings in classes; know how to create an object; know how to define methods; know how to update state through methods; know how to define `let` and `val` bindings in a class; know what static bindings are and how to define them; know about the `unit` type.
- `visibility-basic`: know about the `public` and `private` visibility modifiers.

## Prerequisites

This exercise's prerequisites Concepts are:

- `records-basic`: know what records are.
- `bindings-mutable`: know how to work with mutable bindings.
- `strings-basic`: strings are used as data.
- `integers-basic`: integers are used as data.

### Hints

- [Classes][classes]: introduction to classes.
- [Constructors][constructors]: how to define primary and additional constructors, and how to create objects by calling a constructor.
- [Methods][methods]: how to define methods in classes.
- [`let` bindings in classes][let-bindings-in-classes]: how to use `let` bindings to define fields.
- [`do` bindings in classes][do-bindings-in-classes]: how to use `do` bindings to add constructor logic.
- [`val` keyword][val-keyword]: describes `val` fields.
- [`unit` ][unit-type]: describes the `unit` type.

### After

- [Classes][classes]: introduction to classes.
- [Constructors][constructors]: how to define primary and additional constructors, and how to create objects by calling a constructor.
- [Methods][methods]: how to define methods in classes.
- [`let` bindings in classes][let-bindings-in-classes]: how to use `let` bindings to define fields.
- [`do` bindings in classes][do-bindings-in-classes]: how to use `do` bindings to add constructor logic.
- [`val` keyword][val-keyword]: describes `val` fields.
- [`unit` ][unit-type]: describes the `unit` type.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific logic to be added to the [analyzer][analyzer].

## Implementing

To implement the concept exercise, the following files must be created:

<pre>
languages
└── fsharp
    └── exercises
        └── concept
            └── classes
                ├── .docs
                |   ├── instructions.md
                |   ├── introduction.md
                |   ├── hints.md
                |   └── after.md
                ├── .meta
                |   ├── config.json
                |   |── design.md
                |   └── Example.fs
                ├── Classes.fs
                ├── Classes.fsproj
                └── ClassesTests.fs
</pre>

## Step 1: add .docs/introduction.md

This file contains an introduction to the concept. It should be explicit about what the student should learn from the exercise, and provide a short, concise introduction to the concept(s). The aim is to give the student just enough context to figure things out themselves and solve the exercise, as research has shown that self-discovery is the most effective learning experience. Mentioning technical terms that the student can Google if they so want, is preferable over including any code samples or an extensive description. For example we might describe a string as a "Sequence of Unicode characters" or a "series of bytes" or "an object". Unless the student needs to understand the details of what those mean to be able to solve the exercise we should not give more info in this introduction - instead allowing the student to Google, ignore, or map their existing knowledge.

## Step 2: add .docs/instructions.md

This file contains instructions for the exercise. It should explicitly explain what the student needs to do (define a method with the signature `X(...)` that takes an A and returns a Z), and provide at least one example usage of that function. If there are multiple tasks within the exercise, it should provide an example of each.

## Step 3: add .docs/hints.md

If the student gets stuck, we will allow them to click a button requesting a hint, which shows this file. This will not be a "recommended" path and we will (softly) discourage them using it unless they can't progress without it. As such, it's worth considering that the student reading it will be a little confused/overwhelmed and maybe frustrated.

The file should contain both general and task-specific "hints". These hints should be enough to unblock almost any student. They might link to the docs of the functions that need to be used.

The hints should not spell out the solution, but instead point to a resource describing the solution (e.g. linking to documentation for the function to use).

## Step 4: add .docs/after.md

Once the student completes the exercise they will be shown this file, which should provide them with a summary of what the exercise aimed to teach. This document can also link to any additional resources that might be interesting to the student in the context of the exercise.

These files are also all described in the [concept exercises document][docs-concept-exercises].

## Step 5: update languages/fsharp/config.json

An entry should be added to the track's `config.json` file for the new concept exercise:

```json
{
  ...
  "exercises": {
    "concept": [
      ...
      {
        "slug": "classes",
        "uuid": "3c7fa1a8-ac35-4e9c-956e-d4555f07b970",
        "concepts": ["classes-basic", "visibility-basic"],
        "prerequisites": [
          "records-basic",
          "bindings-mutable",
          "integers-basic",
          "booleans-basic"
        ]
      }
    ]
  }
}
```

## Step 6: adding track-specific files

These files are specific to the F# track:

- `Classes.fs`. the stub implementation file, which is the starting point for students to work on the exercise.
- `Classes.fsproj`: the F# project file.
- `ClassesTests.fs`: the test suite.
- `.meta/Example.fs`: an example implementation that passes all the tests. The project file should _not_ include this file.

## Step 7: update the general concept document

Add the exercise to the [concept's shared document's][reference] `## Implementations` section ([example](https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations)). Note that there currently is no such document, so it has to be created first.

## Step 8: updating list of implemented exercises

- Add the exercise to the [list of implemented exercises][implemented-exercises].

## Step 9: add .meta/design.md:

This file contains information on the exercise's design, which includes things like its goal, its teaching goals, what not to teach, and more ([example][meta-design]). This information can be extracted from this GitHub issue.

## Step 10: add .meta/config.json:

This file contains meta information on the exercise, which currently only includes the exercise's contributors ([example][meta-config.json]).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/fsharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/reference/concepts/classes.md
[analyzer]: https://github.com/exercism/fsharp-analyzer
[representer]: https://github.com/exercism/fsharp-representer
[meta-design]: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/numbers/.meta/design.md
[meta-config.json]: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/numbers/.meta/config.json
[docs-concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[classes]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/classes
[constructors]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/members/constructors
[let-bindings-in-classes]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/members/let-bindings-in-classes
[do-bindings-in-classes]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/members/do-bindings-in-classes
[methods]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/members/methods
[val-keyword]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/members/explicit-fields-the-val-keyword
[unit-type]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/unit-type
